include/ibamr/HierarchyTimeInterpolator.h

```diff
@@ -0,0 +1,182 @@
+ * intervals. This class can also be used to determine intermediate velocity fields when at a periodic steady state.
+ */
+class HierarchyTimeInterpolator
+{
+public:
+    /*!
+     * The constructor for class HierarchyTimeInterpolator sets some default values and determines data centering. The
+     * expected period and number of snapshots must be set in the input database. This class assumes the snapshots will
+     * be taken at equidistant points along the periodic interval.
+     */
+    HierarchyTimeInterpolator(std::string object_name,
+                              SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
+                              SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > var,
+                              SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy);
+
+    /*!
+     * The constructor for class HierarchyTimeInterpolator sets some default values and determines data centering. In
+     * this constructor, the times at which snapshots are taken are set, as well as the period length.
+     */
+    HierarchyTimeInterpolator(std::string object_name,
+                              std::set<double> snapshot_time_points,
+                              double period,
+                              SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > var,
+                              SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy);
+
+    /*!
+     * \brief The destructor for class HierarchyTimeInterpolator deallocates patch data as needed.
+     */
+    ~HierarchyTimeInterpolator();
+
+    /*!
+     * Clear the snapshots taken with this class.
+     */
+    void clearSnapshots();
+
+    /*!
+     * Fill a patch index at a particular time point. This will perform a time interpolation of the snapshots stored in
+     * the SnapshotCache object. Note if there is only one snapshot (i.e. the period is 0 and we are assumed to be at a
+     * steady state), a copy will be performed
+     */
+    void
+    fillSnapshotAtTime(int u_idx, double time, SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy);
+
+    /*!
+     * Update the time averaged snapshot. The update is computed through the relation
+     * ``` u_avg = u_avg + 1/N * (u - u_avg)```
+     * in which N is the number of updates to the snapshot. If this the first update after clearMean() is called, we
+     * simply copy data.
+     *
+     * This function returns true if the snapshot is at a steady state, or that ||u - u_avg|| < threshold.
+     */
+    bool updateTimeAveragedSnapshot(int u_idx,
+                                    double time,
+                                    SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy);
+
+    /*!
+     * Return true if all the tracked mean fields are at a steady state.
+     */
+    bool isAtPeriodicSteadyState();
+
+    /*!
```

Comment:
I had originally set this up so that the number of snapshots and time points must be set in the constructor. I think we can get rid of that requirement, as long as the time points that snapshots are stored aren't too close together.